drivers/adc/adc_mcp320x.c

```diff
@@ -0,0 +1,375 @@
+	int err;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	err = mcp320x_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static int mcp320x_read(struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	return mcp320x_read_async(dev, sequence, NULL);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct mcp320x_data *data = CONTAINER_OF(ctx, struct mcp320x_data, ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct mcp320x_data *data = CONTAINER_OF(ctx, struct mcp320x_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int mcp320x_read_channel(struct device *dev, u8_t channel, u16_t *result)
+{
+	const struct mcp320x_config *config = dev->config->config_info;
+	struct mcp320x_data *data = dev->driver_data;
+	u8_t tx_bytes[2];
+	u8_t rx_bytes[2];
+	int err;
+	const struct spi_buf tx_buf[2] = {
+		{
+			.buf = tx_bytes,
+			.len = sizeof(tx_bytes)
+		},
+		{
+			.buf = NULL,
+			.len = 1
+		}
+	};
+	const struct spi_buf rx_buf[2] = {
+		{
+			.buf = NULL,
+			.len = 1
+		},
+		{
+			.buf = rx_bytes,
+			.len = sizeof(rx_bytes)
+		}
+	};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf)
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf)
+	};
+
+	/*
+	 * Configuration bits consists of: 5 dummy bits + start bit +
+	 * SGL/#DIFF bit + D2 + D1 + D0 + 6 dummy bits
+	 */
+	tx_bytes[0] = BIT(2) | channel >> 2;
+	tx_bytes[1] = channel << 6;
+
+	if ((data->differential & BIT(channel)) == 0) {
+		tx_bytes[0] |= BIT(1);
+	}
+
+	err = spi_transceive(data->spi_dev, &config->spi_cfg, &tx, &rx);
+	if (err) {
+		return err;
+	}
+
+	*result = sys_get_be16(rx_bytes);
+	*result &= BIT_MASK(MCP320X_RESOLUTION);
+
+	return 0;
+}
+
+static void mcp320x_acquisition_thread(struct device *dev)
+{
+	struct mcp320x_data *data = dev->driver_data;
+	u16_t result = 0;
+	u8_t channel;
+	int err;
+
+	while (true) {
+		k_sem_take(&data->sem, K_FOREVER);
+
+		while (data->channels) {
+			channel = find_lsb_set(data->channels) - 1;
+
+			LOG_DBG("reading channel %d", channel);
+
+			err = mcp320x_read_channel(dev, channel, &result);
+			if (err) {
+				LOG_ERR("failed to read channel %d (err %d)",
+					channel, err);
+				adc_context_complete(&data->ctx, err);
+				break;
+			}
+
+			LOG_DBG("read channel %d, result = %d", channel,
+				result);
+
+			*data->buffer++ = result;
+			WRITE_BIT(data->channels, channel, 0);
+		}
+
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static int mcp320x_init(struct device *dev)
+{
+	const struct mcp320x_config *config = dev->config->config_info;
+	struct mcp320x_data *data = dev->driver_data;
+
+	data->dev = dev;
```

Comment:
Is this actually needed? I couldn't find a place where this field was used.

Reply:
No, that was leftover from an earlier implementation. Fixed.
Thank you for reviewing.